eth/protocols/eth/handlers.go

```diff
@@ -274,6 +285,62 @@ func ServiceGetReceiptsQuery(chain *core.BlockChain, query GetReceiptsRequest) [
 	return receipts
 }
 
+// serviceGetReceiptsQuery69 assembles the response to a receipt query.
+// It does not send the bloom filters for the receipts
+func serviceGetReceiptsQuery69(chain *core.BlockChain, query GetReceiptsRequest) []rlp.RawValue {
+	// Gather state data until the fetch or network limits is reached
+	var (
+		bytes    int
+		receipts []rlp.RawValue
+	)
+	for lookups, hash := range query {
+		if bytes >= softResponseLimit || len(receipts) >= maxReceiptsServe ||
+			lookups >= 2*maxReceiptsServe {
+			break
+		}
+		// Retrieve the requested block's receipts
+		results := chain.GetRawReceiptsByHash(hash)
+		if results == nil {
+			if header := chain.GetHeaderByHash(hash); header == nil || header.ReceiptHash != types.EmptyRootHash {
+				continue
+			}
+		} else {
+			block := chain.GetBlockByHash(hash)
```

Comment:
`chain.GetBlockByHash` 

> GetBlockByHash retrieves a block from the database by hash, **caching it if found.**

(emphasis mine). Are we sure we want to do that?  I mean, maybe, I don't know, but worth considering.

Reply:
Also, there are basically two paths here: 

1. The header/block has no txs / recipts
2. The header/block has txs / receipts
3. (error-path: the header/block has it, but we're missing it)

For the two paths, we can distinguish between them if we first read the header, and check the receipts-root. The current code unconditionally tries to load the receipts, which reads header number and  then loads from ancients or regular-database. And then, it always needs to read the header anyway (either explicitly or implicitly when loading the body) to distinguish between 1 or 2. 

So if you read the header first, you can make path 1 less loady.